### ISSUE TITLE
Re using ac-gan code -  Some clarifications. 

### ISSUE THREAD

**Issue by rjpg** [eriklindernoren/Keras-GAN#63]
First of all great work !

Iam seeing the AC-GAN example. I would like to confirm some things if possible :-) 

1 - Every epoch you run only one batch of random examples taken from the set right ? 

2- The output is (e.g.):
453 [D loss: 2.150578, acc.: 51.56%, op_acc: 54.69%] [G loss: 3.275106]
"op_acc" is the accuracy on the second layer output of the descriminator, that is guessing the label right ? 
"acc." is the accuracy on the first layer output of the discriminator, that is guessing if it is fake or real correct ? 

3- on the softmax layer in the discriminator (second layer output - the labels layer ...) you have : 
`label = Dense(self.num_classes+1, activation="softmax")(features)
`why  num_classes+1 ? 

Thanks in advance, great work !

**Comment by rjpg** [comment]
Ok I think I understand point 3 (in my questions): 

> 3- on the softmax layer in the discriminator (second layer output - the labels layer ...) you have :
> label = Dense(self.num_classes+1, activation="softmax")(features)why num_classes+1 ?
 
you have on training  : 
            img_labels = y_train[idx]
            fake_labels = 10 * np.ones(img_labels.shape)

            # Train the discriminator
            d_loss_real = self.discriminator.train_on_batch(imgs, [valid, img_labels])
            d_loss_fake = self.discriminator.train_on_batch(gen_imgs, [fake, fake_labels])
            d_loss = 0.5 * np.add(d_loss_real, d_loss_fake)

So you on fake images you dont use the labels you say is class 10 (fake on softmax..).

It would not be better if you use the label you use on the input of the generator among with the noise ? 
From the previous lines, use the sampled_labels (between 0-9) ? : 
```

            sampled_labels = np.random.randint(0, 10, (batch_size, 1))

            # Generate a half batch of new images
            gen_imgs = self.generator.predict([noise, sampled_labels])  

```
Any reason against it ?

**Comment by rjpg** [comment]
OK for point 2 : 

> 2- The output is (e.g.):
> 453 [D loss: 2.150578, acc.: 51.56%, op_acc: 54.69%] [G loss: 3.275106]
> "op_acc" is the accuracy on the second layer output of the descriminator, that is guessing the label right ?
> "acc." is the accuracy on the first layer output of the discriminator, that is guessing if it is fake or real correct ?

I have the answer. I have put names on last two parallel layers on discriminator : 
        ```
# Determine validity and label of the image
        validity = Dense(1, activation="sigmoid", name="sigmoidSide")(features)
        label = Dense(self.num_classes, activation="softmax",name="softmaxSide")(features)
```

And print them on the output : 
```

0 [D loss: 3.453677, acc.: 42.19%, op_acc: 7.81%] [G loss: 3.525491]
['loss', 'sigmoidSide_loss', 'softmaxSide_loss', 'sigmoidSide_acc', 'softmaxSide_acc']
```

with this code : 
```
print ("%d [D loss: %f, acc.: %.2f%%, op_acc: %.2f%%] [G loss: %f]" % (epoch, d_loss[0], 100*d_loss[3], 100*d_loss[4], g_loss[0]))
            print (self.discriminator.metrics_names)
```

So "acc." is index 3 =  'sigmoidSide_acc' 
and "op_acc" is index 4  =  'softmaxSide_acc'

**Comment by rjpg** [comment]
I am testing using softmax with 10 units (= n_labels) and giving the random labels used in the generator to the discriminator to use as label of the image instead of "fake label" on softmax (for that you have the sigmoid layer...). 

The construction of images does not seam so good (maybe it is my imprecision) but the accuracy in the softmax side seams to converge faster. I want to use this for classification (the generation is not important for me) so : good for me :-) . 

I think gans for classification will be good so I can detect ambiguous or near noise inputs (?)

**Comment by eriklindernoren** [comment]
Sorry, for the late response. You seem to have answered your own questions. :) I have merged the change you suggested in point 3.
